ros/src/sensing/filters/packages/points_preprocessor/nodes/ground_filter/ground_filter.cpp

```diff
@@ -1,128 +1,343 @@
+							} else {
+								unknown_index[unknown_index_size] = index;
+								unknown_index_size++;
+							}
+						}
+						point_index_size = 0;
+					}
+				}
+			}
+			if (j == 0)
+			{
+				if (point_index_size != 0)
+				{
+					if (point_index_size > min_point_)
+					{
+						for (int m = 0; m < point_index_size; m++)
+						{
+							int index = index_map_.at<int>(point_index[m],i);
+							point.x = in_cloud_msg->points[index].x;
+							point.y = in_cloud_msg->points[index].y;
+							point.z = in_cloud_msg->points[index].z;
+							point.intensity = in_cloud_msg->points[index].intensity;
+							point.ring = in_cloud_msg->points[index].ring;
+							out_ground_points.push_back(point);
+							point_class[point_index[m]] = GROUND;
+						}
+						point_index_size = 0;
+					}
+					else
+					{
+						for (int m = 0; m < point_index_size; m++)
+						{
+							int index = index_map_.at<int>(point_index[m],i);
+							point.z = in_cloud_msg->points[index].z;
+							if (point.z > clipping_thres_ - sensor_height_)
+							{
+								point.x = in_cloud_msg->points[index].x;
+								point.y = in_cloud_msg->points[index].y;
+								point.intensity = in_cloud_msg->points[index].intensity;
+								point.ring = in_cloud_msg->points[index].ring;
+								out_groundless_points.push_back(point);
+								point_class[point_index[m]] = VERTICAL;
+							}
+							else
+							{
+								unknown_index[unknown_index_size] = index;
+								unknown_index_size++;
+							}
+						}
+						point_index_size = 0;
+					}//end else
+				}//end if (point_index_size != 0)
+
+				double centroid = 0;
+				int cluster_index[vertical_res_];
+				int cluster_index_size = 0;
+				for (int m = unknown_index_size - 1; m >= 0; m--)
+				{
+					double x0 = in_cloud_msg->points[unknown_index[m]].x;
+					double y0 = in_cloud_msg->points[unknown_index[m]].y;
+					double r0 = sqrt(x0*x0 + y0*y0);
+					double r_diff = fabs(r0 - centroid);
+					if ((r_diff < point_distance_) || cluster_index_size == 0)
+					{
+						cluster_index[cluster_index_size] = unknown_index[m];
+						cluster_index_size++;
+						centroid = r0;
+						if (m == 0)
+						{
+							if(cluster_index_size > 1)
+							{
+								PublishPointCloud(in_cloud_msg, cluster_index, cluster_index_size, out_groundless_points);
+							}
+							else
+							{
+								PublishPointCloud(in_cloud_msg, cluster_index, cluster_index_size, out_ground_points);
+							}
+						}
+					}
+					else
+					{
+						if(cluster_index_size > 1)
+						{
+							PublishPointCloud(in_cloud_msg, cluster_index, cluster_index_size, out_groundless_points);
+						}
+						else
+						{
+							PublishPointCloud(in_cloud_msg, cluster_index, cluster_index_size, out_ground_points);
+						}
+					}
+				}//end for (int m = unknown_index_size - 1; m >= 0; m--)
+			}//end if (j == 0)
+		}
+	}
 
-	RemoveFloor(current_sensor_cloud_ptr, lanes_cloud_ptr, ground_cloud_ptr, points_distance_, angle_threshold_);
+}
```

Comment:
@amc-nu @dejanpan  apologies for slow response, I don't have full documentation explaining the algorithm yet. However, I do have some PowerPoint slides which explain main idea of the algorithm. In case you would like to take a look, I uploaded it to the following link https://drive.google.com/file/d/0B8iMqYxCmvZAZGRVOUhVQmhsSVE/view?usp=sharing